File: spark/src/main/scala/geotrellis/spark/io/FilteringLayerReader.scala
package geotrellis.spark.io

import geotrellis.spark._
import geotrellis.spark.io.avro._
import geotrellis.spark.io.json._

import org.apache.spark.rdd._
import spray.json._

import scala.reflect._

abstract class FilteringLayerReader[ID] extends LayerReader[ID] {

  def read[
    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag,
    V: AvroRecordCodec: ClassTag,
    M: JsonFormat
  ](id: ID, rasterQuery: RDDQuery[K, M], numPartitions: Int): RDD[(K, V)] with Metadata[M]

  def read[
    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag,
    V: AvroRecordCodec: ClassTag,
    M: JsonFormat
  ](id: ID, rasterQuery: RDDQuery[K, M]): RDD[(K, V)] with Metadata[M] =
    read(id, rasterQuery, defaultNumPartitions)

  def read[
    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag,
    V: AvroRecordCodec: ClassTag,
    M: JsonFormat
  ](id: ID, numPartitions: Int): RDD[(K, V)] with Metadata[M] =
    read(id, new RDDQuery[K, M], numPartitions)

  def query[
    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag,
    V: AvroRecordCodec: ClassTag,
    M: JsonFormat
  ](layerId: ID): BoundRDDQuery[K, M, RDD[(K, V)] with Metadata[M]] =
    new BoundRDDQuery(new RDDQuery, read(layerId, _))

  def query[
    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag,
    V: AvroRecordCodec: ClassTag,
    M: JsonFormat
  ](layerId: ID, numPartitions: Int): BoundRDDQuery[K, M, RDD[(K, V)] with Metadata[M]] =
    new BoundRDDQuery(new RDDQuery, read(layerId, _, numPartitions))
}

File: geotrellis-python/geotrellis/spark/io/FilteringLayerReader.py
from __future__ import absolute_import
from geotrellis.spark.io.LayerQuery import LayerQuery, BoundLayerQuery
from geotrellis.spark.io.LayerReader import LayerReader

class FilteringLayerReader(LayerReader):

    def _read1(self, K, V, M, _id, rasterQuery):
        return self._read2(K, V, M, _id, rasterQuery, self.defaultNumPartitions)

    def _read2(self, K, V, M, _id, rasterQuery, numPartitions):
        indexFilterOnly = False
        return self._read3(K, V, M, _id, rasterQuery, numPartitions, indexFilterOnly)

    def _read3(self, K, V, M, _id, rasterQuery, numPartitions, indexFilterOnly):
        pass

    def _read_id_num(self, K, V, M, _id, numPartitions):
        return self._read2(K, V, M, _id, LayerQuery(), numPartitions)

    def query(self, K, V, M, layerid, numPartitions = None):
        if numPartitions is None:
            return BoundLayerQuery(LayerQuery(), lambda x: self.read(K, V, M, layerid, x))
        else:
            return BoundLayerQuery(LayerQuery(), lambda x: self.read(K, V, M, layerid, x, numPartitions))

